extensions/amp-bind/amp-bind.md

```diff
@@ -66,425 +72,1074 @@ The `amp-bind` component allows you to add custom stateful interactivity to your
-the previous value of variables set by `AMP.pushState()`.
+Nested objects are generally merged to a maximum depth of 10. All variables, including those defined in `<amp-state>`, can be overidden.
 
-For example:
+[example preview="inline" playground="true" imports="amp-bind"]
 
 ```html
-<button on="tap:AMP.pushState({foo: '123'})">Set 'foo' to 123</button>
+<amp-state id="myState">
+  <script type="application/json">
+    {
+      "foo": "bar",
+      "first": {
+        "a": "nested once",
+        "ab": {
+          "b": "nested twice",
+          "bc": {
+            "c": "nested three times",
+            "cd": {
+              "d": "nested four times",
+              "de": {
+                "e": "nested five times",
+                "ef": {
+                  "f": "nested six times",
+                  "fg": {
+                    "g": "nested seven times",
+                    "gh": {
+                      "h": "nested nine times",
+                      "hi": {
+                        "i": "nested ten times"
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  </script>
+</amp-state>
+<p [text]="myState.foo"></p>
+<p [text]="myState.first.ab.bc.cd.de.ef.fg.gh.hi.i"></p>
+<button on="tap:AMP.setState({})">See value of set state</button>
+<button
+  on="tap:AMP.setState({ myState: {first: {ab: {bc: {cd: {de: {ef: {fg: {gh: {hi: {i:'this is as far as you should merge nested values'} } } } } } } } } } })"
+>
+  Merge 10th nested object
+</button>
 ```
 
-- Tapping the button will set variable `foo` to 123 and push a new history entry.
-- Navigating back will restore `foo` to its previous value, "bar" (equivalent to calling `AMP.setState({foo: 'bar'})`.
+[/example]
+
+#### Circular references
+
+`AMP.setState(object)` throws an error if `object` contains a circular reference.
 
-### Expressions
+#### Removing a variable
 
-Expressions are similar to JavaScript with some important differences.
+Remove an existing state variable by setting its value to `null` in `AMP.setState()`.
 
-#### Differences from JavaScript
+```html
+<button on="tap:AMP.setState({removeMe: null})"></button>
+```
 
-- Expressions may only access the containing document's [state](#state).
-- Expressions **do not** have access to `window` or `document`. `global` references the top-level state.
-- Only [white-listed functions](#white-listed-functions) and operators may be used. Custom functions, classes and loops are disallowed. Arrow functions are allowed as function parameters e.g. `[1, 2, 3].map(x => x + 1)`.
-- Undefined variables and array-index-out-of-bounds return `null` instead of `undefined` or throwing errors.
-- A single expression is currently capped at 50 operands for performance. Please [contact us](https://github.com/ampproject/amphtml/issues/new) if this is insufficient for your use case.
+#### Deep-merge with `AMP.setState()`
```

Comment:
Hasn't this been explained before?

Reply:
Sort of. It doesn't go into detail about re-writing the JSON object...